html/modules/custom/docstore/src/ManageFields.php

```diff
@@ -125,80 +136,176 @@ protected function machineNameExists($machine_name, $entity_type) {
+  }
+
   /**
    * Add field to index.
+   *
+   * Note: additional fields ends with a `_` to avoid clash with the machine
+   * names of user created fields. This is indeed a forbidden pattern.
+   * We use `_` as suffix and not prefix so that sorting a list of fields by
+   * field names will ensure the extra fields appear after the base fields,
+   * easing their processing.
+   *
+   * @param string $index_name
+   *   Search API index name.
+   * @param string $datasource_id
+   *   Source of the data for the field. For example `entity:node` for the a
+   *   node field.
+   * @param \Drupal\field\Entity\FieldConfig $field_config
+   *   Field config.
+   * @param string $label
+   *   Field label.
    */
-  public function addDocumentFieldToIndex($field_name, $field_type, $label) {
-    $field_type_mapping = $this->allowedFieldTypes();
+  public function addFieldToIndex($index_name, $datasource_id, FieldConfig $field_config, $label) {
+    $field_name = $field_config->getName();
+    $field_type = $field_config->getType();
 
     // Skip unknown field types.
+    $field_type_mapping = $this->allowedFieldTypes();
     if (!isset($field_type_mapping[$field_type])) {
       return;
     }
 
-    $index = Index::load('documents');
+    // Load the search api index.
+    $index = Index::load($index_name);
+    if (empty($index)) {
+      return;
+    }
 
     // Skip existing fields.
     if ($index->getField($field_name)) {
       return;
     }
 
-    // Add node title if needed.
-    if ($field_type === 'geofield') {
-      $field = new Field($index, $field_name . '_latlon');
-      $field->setType('string');
-      $field->setPropertyPath($field_name . ':latlon');
-      $field->setDatasourceId('entity:node');
-      $field->setLabel($label . ' (latlon)');
-      $index->addField($field);
-    }
-    else {
+    // Add the base field.
+    if ($field_type !== 'geofield') {
       $field = new Field($index, $field_name);
       $field->setType($field_type_mapping[$field_type]);
       $field->setPropertyPath($field_name);
-      $field->setDatasourceId('entity:node');
+      $field->setDatasourceId($datasource_id);
       $field->setLabel($label);
       $index->addField($field);
     }
-
-    // Add node title if needed.
-    if ($field_type === 'node_reference') {
-      $field = new Field($index, $field_name . '_label');
-      $field->setType('string');
-      $field->setPropertyPath($field_name . ':entity:title');
-      $field->setDatasourceId('entity:node');
-      $field->setLabel($label . ' (title)');
-      $index->addField($field);
-    }
-
-    // Add term name if needed.
-    if ($field_type === 'term_reference') {
-      $field = new Field($index, $field_name . '_label');
+    // For geofield fields we index the latlon property which is a string
+    // with the latitude and longitude separated by a comma.
+    // @todo review if that's the most appropriate wau to store a geofield.
```

Comment:
Something to discuss, do we want searches based on location as well?

Reply:
Created https://humanitarian.atlassian.net/browse/DOC-51 to discuss it.